## Importación de las librerías necesarias

In [140]:
import pandas as pd
import numpy as np

## Carga de datos y exploración del respectivo dataframe

In [141]:
df_movie=pd.read_csv(r"Datasets/movies_dataset.csv",delimiter=",")

In [142]:
# Visualización de las doce primeras columnas
df_movie.iloc[:3,:12]

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   

     original_title                                           overview  \
0         Toy Story  Led by Woody, Andy's toys live happily in his ...   
1           Jumanji  When siblings Judy and Peter discover an encha...   
2  Grumpier Old Men  A family wedding reignites the ancient feud be...   

   popularity                       poster_path  
0   21.946943  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg  
1   17.015539  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg  
2   11.712900  /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg

In [143]:
# Visualización de las últimas doce columnas
df_movie.iloc[:3,12:]

production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2  [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   

                                production_countries release_date    revenue  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-10-30  373554033   
1  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-15  262797249   
2  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22          0   

   runtime                                   spoken_languages    status  \
0     81.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
1    104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   
2    101.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   

                                             tagline             title  video  \
0                                                NaN         Toy Story  False   
1          Roll the dice and unleash the excitement!           Jumanji  False   
2  Still Yelling. Still Fighting. Still Ready for...  Grumpier Old Men  False   

   vote_average  vote_count  
0           7.7        5415  
1           6.9        2413  
2           6.5          92

In [144]:
# Información del data frame
df_movie.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45463 entries, 0 to 45462
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  45463 non-null  bool   
 1   belongs_to_collection  4491 non-null   object 
 2   budget                 45463 non-null  int64  
 3   genres                 45463 non-null  object 
 4   homepage               7779 non-null   object 
 5   id                     45463 non-null  int64  
 6   imdb_id                45446 non-null  object 
 7   original_language      45452 non-null  object 
 8   original_title         45463 non-null  object 
 9   overview               44509 non-null  object 
 10  popularity             45463 non-null  float64
 11  poster_path            45080 non-null  object 
 12  production_companies   45463 non-null  object 
 13  production_countries   45463 non-null  object 
 14  release_date           45379 non-null  object 
 15  re

* #### Observación: se procederá a eliminar las filas duplicadas en las cuales los valores de todas las columnas sean idénticos. Esto se realiza con el fin de garantizar la integridad y consistencia de los datos en el data frame. 

In [145]:
# Observando los valores duplicados
duplicates=df_movie[df_movie.duplicated(keep=False)].sort_values(by="id")
print(duplicates.shape[0])
duplicates.head(8)

33


adult belongs_to_collection  budget  \
7345   False                   NaN       0   
9165   False                   NaN       0   
24843  False                   NaN       0   
14012  False                   NaN       0   
22150  False                   NaN       0   
14000  False                   NaN       0   
8068   False                   NaN       0   
9327   False                   NaN       0   

                                                  genres  \
7345   [{'id': 80, 'name': 'Crime'}, {'id': 18, 'name...   
9165   [{'id': 80, 'name': 'Crime'}, {'id': 18, 'name...   
24843  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
14012  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
22150  [{'id': 28, 'name': 'Action'}, {'id': 27, 'nam...   
14000  [{'id': 28, 'name': 'Action'}, {'id': 27, 'nam...   
8068   [{'id': 12, 'name': 'Adventure'}, {'id': 16, '...   
9327   [{'id': 12, 'name': 'Adventure'}, {'id': 16, '...   

                                     homepage     id    imdb_id  \
7345                                      NaN   5511  tt0062229   
9165                                      NaN   5511  tt0062229   
24843            http://www.dealthemovie.com/  11115  tt0446676   
14012            http://www.dealthemovie.com/  11115  tt0446676   
22150  http://www.daysofdarknessthemovie.com/  18440  tt0499456   
14000  http://www.daysofdarknessthemovie.com/  18440  tt0499456   
8068                                      NaN  23305  tt0295682   
9327                                      NaN  23305  tt0295682   

      original_language    original_title  \
7345                 fr       Le Samouraï   
9165                 fr       Le Samouraï   
24843                en              Deal   
14012                en              Deal   
22150                en  Days of Darkness   
14000                en  Days of Darkness   
8068                 en       The Warrior   
9327                 en       The Warrior   

                                                overview  ...  release_date  \
7345   Hitman Jef Costello is a perfectionist who alw...  ...    1967-10-25   
9165   Hitman Jef Costello is a perfectionist who alw...  ...    1967-10-25   
24843  As an ex-gambler teaches a hot-shot college ki...  ...    2008-01-29   
14012  As an ex-gambler teaches a hot-shot college ki...  ...    2008-01-29   
22150  When a comet strikes Earth and kicks up a clou...  ...    2007-01-01   
14000  When a comet strikes Earth and kicks up a clou...  ...    2007-01-01   
8068   In feudal India, a warrior (Khan) who renounce...  ...    2001-09-23   
9327   In feudal India, a warrior (Khan) who renounce...  ...    2001-09-23   

      revenue runtime                           spoken_languages    status  \
7345    39481   105.0  [{'iso_639_1': 'fr', 'name': 'Français'}]  Released   
9165    39481   105.0  [{'iso_639_1': 'fr', 'name': 'Français'}]  Released   
24843       0    85.0   [{'iso_639_1': 'en', 'name': 'English'}]  Released   
14012       0    85.0   [{'iso_639_1': 'en', 'name': 'English'}]  Released   
22150       0    89.0   [{'iso_639_1': 'en', 'name': 'English'}]  Released   
14000       0    89.0   [{'iso_639_1': 'en', 'name': 'English'}]  Released   
8068        0    86.0    [{'iso_639_1': 'hi', 'name': 'हिन्दी'}]  Released   
9327        0    86.0    [{'iso_639_1': 'hi', 'name': 'हिन्दी'}]  Released   

                                                 tagline             title  \
7345   There is no solitude greater than that of the ...       Le Samouraï   
9165   There is no solitude greater than that of the ...       Le Samouraï   
24843                                                NaN              Deal   
14012                                                NaN              Deal   
22150                                                NaN  Days of Darkness   
14000                                                NaN  Days of Darkness   
8068                                                 NaN       The Warrior   
93

In [146]:
# Eliminando los duplicados
df_movie.drop_duplicates(inplace=True)
df_movie=df_movie.reset_index(drop=True)

## Transformaciones

#### 1) Desanidación de los campos anidados que se encuentran en forma de diccionarios

In [147]:
# Columnas en forma de diccionario
df_movie.iloc[:3,[1,3,12,13,17]]

belongs_to_collection  \
0  {'id': 10194, 'name': 'Toy Story Collection', ...   
1                                                NaN   
2  {'id': 119050, 'name': 'Grumpy Old Men Collect...   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2  [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   

                                production_countries  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...   
1  [{'iso_3166_1': 'US', 'name': 'United States o...   
2  [{'iso_3166_1': 'US', 'name': 'United States o...   

                                    spoken_languages  
0           [{'iso_639_1': 'en', 'name': 'English'}]  
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  
2           [{'iso_639_1': 'en', 'name': 'English'}]

* Columna `belongs_to_collection`

Este campo almacena las rutas relativas de las imágenes de las películas. Sin embargo, estos datos no son necesarios, ya que no se requieren en ninguna etapa del proyecto. Además, algunas de las rutas ya no son funcionales y solo hay un número reducido de valores no nulos (4491). Por lo tanto, solo se considerarán los valores de las claves `name`.

In [148]:
df_movie.belongs_to_collection[0]

"{'id': 10194, 'name': 'Toy Story Collection', 'poster_path': '/7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg', 'backdrop_path': '/9FBwqcd9IRruEDUrTdcaafOMKUq.jpg'}"

In [149]:
# Aplicando la función 'name'
df_movie["belongs_to_collection"]=df_movie.belongs_to_collection.apply(lambda dic: eval(dic)["name"] if pd.notnull(dic) else dic)

* Columnas `genres`, `production_companies`, ``production_countries`` y `spoken_languages`

Se guardarán exclusivamente los valores de las claves ``name`` en una lista, ya que los ID no son necesarios. Al extraer únicamente los valores de la clave `name`, se simplifica y conserva la información más relevante y descriptiva de los campos correspondientes. Esta acción mejora la legibilidad, comprensión y utilidad de los datos en el contexto del análisis y la visualización de los datos.

In [150]:
print(df_movie.genres[0])
print(df_movie.production_companies[1])
print(df_movie.production_countries[6])
print(df_movie.spoken_languages[5])

[{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]
[{'name': 'TriStar Pictures', 'id': 559}, {'name': 'Teitler Film', 'id': 2550}, {'name': 'Interscope Communications', 'id': 10201}]
[{'iso_3166_1': 'DE', 'name': 'Germany'}, {'iso_3166_1': 'US', 'name': 'United States of America'}]
[{'iso_639_1': 'en', 'name': 'English'}, {'iso_639_1': 'es', 'name': 'Español'}]


In [151]:
# Función que extrae los valores de las claves 'name'
def name(string):
    return [dic["name"] for dic in eval(string)]

In [152]:
# Aplicando la función 'name'
df_movie["genres"]=df_movie.genres.apply(name)
df_movie["production_companies"]=df_movie.production_companies.apply(name)
df_movie["production_countries"]=df_movie.production_countries.apply(name)
df_movie["spoken_languages"]=df_movie.spoken_languages.apply(name)

# Visualizando los resultados
df_movie.loc[:7,["belongs_to_collection","genres","production_companies","production_countries","spoken_languages"]]

belongs_to_collection                              genres  \
0            Toy Story Collection         [Animation, Comedy, Family]   
1                             NaN        [Adventure, Fantasy, Family]   
2       Grumpy Old Men Collection                   [Romance, Comedy]   
3                             NaN            [Comedy, Drama, Romance]   
4  Father of the Bride Collection                            [Comedy]   
5                             NaN    [Action, Crime, Drama, Thriller]   
6                             NaN                   [Comedy, Romance]   
7                             NaN  [Action, Adventure, Drama, Family]   

                                production_companies  \
0                          [Pixar Animation Studios]   
1  [TriStar Pictures, Teitler Film, Interscope Co...   
2                     [Warner Bros., Lancaster Gate]   
3           [Twentieth Century Fox Film Corporation]   
4       [Sandollar Productions, Touchstone Pictures]   
5  [Regency Enterprises, Forward Pass, Warner Bros.]   
6  [Paramount Pictures, Scott Rudin Productions, ...   
7                             [Walt Disney Pictures]   

                  production_countries     spoken_languages  
0           [United States of America]            [English]  
1           [United States of America]  [English, Français]  
2           [United States of America]            [English]  
3           [United States of America]            [English]  
4           [United States of America]            [English]  
5           [United States of America]   [English, Español]  
6  [Germany, United States of America]  [Français, English]  
7           [United States of America]   [English, Deutsch]

#### 2) Imputación de los valores nulos de los campos ``revenue`` y ``budget`` por el número 0

No existen valores nulos en estos campos

In [153]:
print(df_movie[df_movie.revenue.isnull()].shape[0])
print(df_movie[df_movie.budget.isnull()].shape[0])

0
0


#### 3) Eliminación de  los valores nulos del campo `release_date`

In [154]:
# Visualizando el porcentaje de valores nulos de la columna
f"{(df_movie[df_movie.release_date.isnull()].shape[0]/df_movie.shape[0])*100}%"

'0.18483474893279936%'

In [158]:
df_movie.dropna(subset="release_date",inplace=True)

#### 4) Conversión de las fechas al formato AAAA-mm-dd y creación de la columna ``release_year`` 

El único campo de fecha disponible se encuentra en la columna ``release_date``, el cual se convertirá al tipo 'date'. A partir de este campo, se extraerá el año de estreno de cada película.

In [160]:
df_movie["release_date"]=pd.to_datetime(df_movie.release_date)
df_movie["release_year"]=df_movie.release_date.dt.year

# Visualizando los resultados
df_movie.loc[:,["release_date","release_year"]]

release_date  release_year
0       1995-10-30          1995
1       1995-12-15          1995
2       1995-12-22          1995
3       1995-12-22          1995
4       1995-02-10          1995
...            ...           ...
45440   1991-05-13          1991
45442   2011-11-17          2011
45443   2003-08-01          2003
45444   1917-10-21          1917
45445   2017-06-09          2017

[45362 rows x 2 columns]

#### 5) Creación de la columna ``return``: ``revenue`` / ``budget``

* Observación: Si el campo ``budget`` es igual a cero, el valor de ``return`` será establecido como cero, ya que la división por cero no está definida y no es posible realizarla.

In [161]:
# La mayoría de películas tienen un presupuesto (budget) de cero
f"{(df_movie[df_movie.budget==0].budget.shape[0]/df_movie.shape[0])*100}%"

'80.4131211145893%'

In [162]:
# Creando la columna 'return'
df_movie["return"]=df_movie.apply(lambda x: round(x.revenue/x.budget,2) if x.budget!=0 else 0,axis=1)
df_movie["return"]

0        12.45
1         4.04
2         0.00
3         5.09
4         0.00
         ...  
45440     0.00
45442     0.00
45443     0.00
45444     0.00
45445     0.00
Name: return, Length: 45362, dtype: float64

#### 6) Eliminación de las columnas que no serán utilizadas: ``video``, ``imdb_id``, ``adult``, ``original_title``, `vote_count`,``poster_path`` y ``homepage``.

In [163]:
df_movie.drop(columns=["video","imdb_id","adult","original_title","vote_count","poster_path","homepage"],inplace=True)

* Visualizando el data frame resultante

In [164]:
df_movie

belongs_to_collection    budget                        genres  \
0                Toy Story Collection  30000000   [Animation, Comedy, Family]   
1                                 NaN  65000000  [Adventure, Fantasy, Family]   
2           Grumpy Old Men Collection         0             [Romance, Comedy]   
3                                 NaN  16000000      [Comedy, Drama, Romance]   
4      Father of the Bride Collection         0                      [Comedy]   
...                               ...       ...                           ...   
45440                             NaN         0      [Drama, Action, Romance]   
45442                             NaN         0                       [Drama]   
45443                             NaN         0     [Action, Drama, Thriller]   
45444                             NaN         0                            []   
45445                             NaN         0                            []   

           id original_language  \
0         862                en   
1        8844                en   
2       15602                en   
3       31357                en   
4       11862                en   
...       ...               ...   
45440   30840                en   
45442  111109                tl   
45443   67758                en   
45444  227506                en   
45445  461257                en   

                                                overview  popularity  \
0      Led by Woody, Andy's toys live happily in his ...   21.946943   
1      When siblings Judy and Peter discover an encha...   17.015539   
2      A family wedding reignites the ancient feud be...   11.712900   
3      Cheated on, mistreated and stepped on, the wom...    3.859495   
4      Just when George Banks has recovered from his ...    8.387519   
...                                                  ...         ...   
45440  Yet another version of the classic epic, with ...    5.683753   
45442  An artist struggles to finish his work while a...    0.178241   
45443  When one of her hits goes wrong, a professiona...    0.903007   
45444  In a small town live two brothers, one a minis...    0.003503   
45445  50 years after decriminalisation of homosexual...    0.163015   

                                    production_companies  \
0                              [Pixar Animation Studios]   
1      [TriStar Pictures, Teitler Film, Interscope Co...   
2                         [Warner Bros., Lancaster Gate]   
3               [Twentieth Century Fox Film Corporation]   
4           [Sandollar Productions, Touchstone Pictures]   
...                                                  ...   
45440  [Westdeutscher Rundfunk (WDR), Working Title F...   
45442                                      [Sine Olivia]   
45443                          [American World Pictures]   
45444                                        [Yermoliev]   
45445                                                 []   

                                    production_countries release_date  \
0                             [United States of America]   1995-10-30   
1                             [United States of America]   1995-12-15   
2                             [United States of America]   1995-12-22   
3                             [United States of America]   1995-12-22   
4                             [United States of America]   1995-02-10   
...                                                  ...          ...   
45440  [Canada, Germany, United Kingdom, United State...   1991-05-13   
45442                                      [Philippines]   2011-11-17   
45443                         [United States of America]   2003-08-01   
45444                                           [Russia]   1917-10-21   
45445                                   [United Kingdom]   2017-06-09   

         revenue  runtime     spoken_languages    status  \
0      373554033     81.0            [English]  Released   
1      262797249    104.0  [English, Fr

In [165]:
df_movie.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45362 entries, 0 to 45445
Data columns (total 19 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   belongs_to_collection  4488 non-null   object        
 1   budget                 45362 non-null  int64         
 2   genres                 45362 non-null  object        
 3   id                     45362 non-null  int64         
 4   original_language      45351 non-null  object        
 5   overview               44421 non-null  object        
 6   popularity             45362 non-null  float64       
 7   production_companies   45362 non-null  object        
 8   production_countries   45362 non-null  object        
 9   release_date           45362 non-null  datetime64[ns]
 10  revenue                45362 non-null  int64         
 11  runtime                45116 non-null  float64       
 12  spoken_languages       45362 non-null  object        
 13  s

### Exportación del data frame "df_platform" en un archivo CSV para la realización de los endpoints

In [166]:
df_movie.to_csv("movie_transformation.csv",index=False)

---